harness/determined/profiler.py

```diff
@@ -0,0 +1,524 @@
+        #       and we have no mechanism to handle the inbound queue filling up
+        max_attempts = 5
+
+        # TODO: Convert to work with new multiple batch API
+        for post_body in post_bodies:
+            for i in range(max_attempts):
+                try:
+                    api.post_trial_profiler_metrics(
+                        self.master_url,
+                        post_body["values"],
+                        post_body["batches"],
+                        post_body["timestamps"],
+                        post_body["labels"],
+                    )
+                    break
+                except Exception as e:
+                    # TODO: We could handle specific API errors differently as some are non-recoverable
+                    # TODO: Log info about error
+                    time.sleep(i**2)  # exponential backoff
+
+
+    def kill(self):
+        self.quitting = True
+
+
+
+
+
+GIGA = 1_000_000_000
+
+class SimpleCpuUtilCollector:
+    def measure(self, batch_idx):
+        cpu_util = psutil.cpu_percent()
+        timestamp = datetime.datetime.utcnow()
+        return Measurement(timestamp, batch_idx, cpu_util)
+
+
+
+class FreeMemoryCollector:
+    # We choose to report free memory instead of available memory because it is useful to
+    # be able to see memory usage for cached files, but we could change to available instead
+    # https://psutil.readthedocs.io/en/latest/#psutil.virtual_memory
+    def measure(self, batch_idx):
+        free_mem_bytes = psutil.virtual_memory().free
+        timestamp = datetime.datetime.utcnow()
+        return Measurement(timestamp, batch_idx, free_mem_bytes * GIGA)
+
+
+
+class NetThroughputCollector:
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.start_time = time.time()
+        net = psutil.net_io_counters()
+        self.start_sent = net.bytes_sent
+        self.start_recv = net.bytes_recv
+
+    def measure(self, batch_idx):
+        net = psutil.net_io_counters()
+        end_time = time.time()
+
+        sent_bytes_delta = net.bytes_sent - self.start_sent
+        recv_bytes_delta = net.bytes_recv - self.start_recv
+
+        time_delta = end_time - self.start_time
+
+        self.start_time = end_time
+        self.start_sent = net.bytes_sent
+        self.start_recv = net.bytes_recv
+
+        sent_throughput_bytes_per_second = sent_bytes_delta / time_delta
+        recv_throughput_bytes_per_second = recv_bytes_delta / time_delta
+
+        sent_throughput_gigabits_per_second = sent_throughput_bytes_per_second * 8 * GIGA
+        recv_throughput_gigabits_per_second = recv_throughput_bytes_per_second * 8 * GIGA
+
+        timestamp = datetime.datetime.fromtimestamp(end_time)
+        return Measurement(timestamp, batch_idx, sent_throughput_gigabits_per_second), \
+               Measurement(timestamp, batch_idx, recv_throughput_gigabits_per_second)
+
+
+class DiskReadWriteRateCollector:
```

Comment:
Collector per metric got pretty verbose it seems. Didn't you look at what yoav did forever ago here https://github.com/determined-ai/determined/blob/master/harness/determined/layers/_harness_profiler.py#L113-L133? Could save a bit of space but not really a show stopper.